### DISCUSSION TITLE
Why does pgsync need to both pull from db and read from the replication logs?

### DISCUSSION THREAD

**Discussion by jeremykhon** [toluaina/pgsync#396] [Q&A]
I'm just reading these two lines from the `pull` function in `pgsync/sync.py` and just wonder why isn't the `self.sync` function enough? why would there be anything missed if `self.sync` will query everything from txmin to txmax? Why do we need to call `self.logical_slot_changes` to sync everything from txmin to txmax again?

```python
def pull(self) -> None:
        """Pull data from db."""
        txmin: int = self.checkpoint
        txmax: int = self.txid_current
        logger.debug(f"pull txmin: {txmin} - txmax: {txmax}")

        # forward pass sync
        self.es.bulk(self.index, self.sync(txmin=txmin, txmax=txmax))

        # now sync up to txmax to capture everything we may have missed
        self.logical_slot_changes(txmin=txmin, txmax=txmax, upto_nchanges=None)

        self.checkpoint: int = txmax or self.txid_current
        self._truncate = True
```
Thank you for working on this package

**Comment by toluaina** [comment]
`logical_slot_changes` is necessary because if pgsync was stopped for some time and during that period, changes were made to the database. 
in this case, there is no way for pgsync to know e.g what deletion operations were made
